nntrainer/layers/acti_func.h

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   acti_func.cpp
+ * @date   22 March 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @author Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is Activation Function Class for Neural Network
+ *
+ */
+
+#ifndef __ACTI_FUNC_H__
+#define __ACTI_FUNC_H__
+#ifdef __cplusplus
+
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @brief     Enumeration of activation function type
+ */
+enum class ActivationType {
+  ACT_TANH,    /** tanh */
+  ACT_SIGMOID, /** sigmoid */
+  ACT_RELU,    /** ReLU */
+  ACT_SOFTMAX, /** softmax */
+  ACT_NONE,    /** no op */
+  ACT_UNKNOWN  /** unknown */
+};
+
+/**
+ * @class   ActiFunc Class
+ * @brief   ActiFunc Class
+ */
+class ActiFunc {
+
+public:
+  /**
+   * @brief     Constructor of ActiFunc
+   */
+  ActiFunc(ActivationType at = ActivationType::ACT_NONE) { setActiFunc(at); }
+
+  /**
+   * @brief     Destructor of ActiFunc
+   */
+  ~ActiFunc(){};
+
+  /**
+   * @brief setActivation by preset ActivationType
+   *
+   * @param[in] ActivationType
+   */
+  void setActiFunc(ActivationType acti_type);
+
+  /**
+   * @brief run function
+   *
+   * @param[in] x : input
+   * @param[out] output : output
+   */
+  void run_fn(Tensor const &x, Tensor &output);
+
+  /**
+   * @brief run prime function
+   *
+   * @param[in] in : input
+   * @param[out] ret : output
+   * @param[in] deriv : derivative
+   * @retVal    Tensor
+   */
+  Tensor &run_prime_fn(Tensor &in, Tensor &ret, Tensor const &deriv);
+
+  /**
+   * @brief       Calculate softmax for Tensor Type
+   * @param[in] x Tensor
+   * @param[out] output output Tensor
+   * @retval      Tensor
+   */
+  static Tensor &softmax(Tensor const &x, Tensor &output);
```

Comment:
If the purpose of this class is to expose an interface to be able to create activation functions and run the forward/backward given input/output, then let's make these static activation functions private, and if any layer wants to use these functions, they must create an object of this class.

Reply:
> If the purpose of this class is to expose an interface to be able to create activation functions and run the forward/backward given input/output, then let's make these static activation functions private, and if any layer wants to use these functions, they must create an object of this class.

I'm not sure about this. The loss layer also use this Activation function.. In order to change to use actifunc object, then we have to make the wrapper of each function. and I don't think it is no different to use these function as public.

Reply:
If we intend to directly use the functions given in this class, then can you please elaborate on the purpose of this class. 
As per my understanding of the PR description, it seems like this class facilitates the use of activation functions in other layers.

Reply:
> , it seems like this class facilitates the use of activation functions in other layers.
It is the purpose of this class. These functions was in ActivationLayer Class and it was public. That's because loss layer use them directly. What I'm saying is it Is too much wrapper code if we make them private and make public code to use directly. I think it is ok as it is now.

Reply:
I agree that using ActiFunc obj with acti_func.h header is a good proposal and makes structure cleaner.
However, if the usage of activation functions is still directly using the ActiFunc obj (like in loss layer), then this class makes little difference.
I would prefer if loss layer can also be updated to have a ActiFunc obj just like RNNLayer will have. And now, if all users of ActiLayer will create its object, then there is no need to open individual activation functions as public.

Reply:
This can be done in another PR then. I will make issue for it.( #1040 )

Reply:
alright.